proposals/p3848.md

```diff
@@ -0,0 +1,638 @@
+                       // mode specifier of `copy`
+  }
+  var thread: Thread = Thread.Make(MyThread);
+  thread.Join();
+}
+```
+
+#### Alternative Considered: Middle vs Beginning
+
+Previously, it was proposed that the default capture mode would come after all
+the explicit captures. In addition, it was proposed that the lack of any square
+brackets opted the function into capturing by default with an implicitly defined
+default capture mode. These behaviors do not mirror lambdas in C++ and so were
+decided against. Primarily, it was recognized that it's valuable to be able to
+intermix explicit captures with deduced parameters and fields in any order that
+makes the most sense for the context. Without a clear justification for a rule
+that says they can't intermix, the more flexible behavior was favored.
+
+## Function Fields
+
+**Proposal**: Function fields mirror the behavior of init captures in C++. A
+function field definition consists of an irrefutable pattern, `=`, and an
+initializer. It matches the pattern with the initializer when the function
+definition is evaluated. The bindings in the pattern have the same lifetime as
+the function, and their scope extends to the end of the function body.
+
+To prevent ambiguities, function fields can only exist on functions where the
+definition is attached to the declaration. This means they are supported on
+lambdas and they are supported on function declarations that are immediately
+defined, but they are not supported on forward-declared functions.
+
+```
+fn Foo() {
+  var h: Handle = Handle.Get();
+  var thread: Thread = Thread.Make(fn [a: auto = h; var b: auto = h] {
+    a.Process();  // ❌ Invalid: Cannot call mutating methods on value bindings
+    b.Process();  // ✅ Valid: Can mutate object bindings
+  });
+  thread.Join();
+}
+```
+
+### Alternative Considered: No Function Fields
+
+Alternatively, by-value and by-object capturing could serve the same purpose.
+This was decided against because capturing is not as expressive as general
+purpose binding patterns. The lack of an initializing expression would create an
+ergonomic burden.
+
+## Copy Semantics
+
+**Proposal**: To mirror the behavior of C++, function declarations and lambdas
+will be as copyable as their contained function fields and function captures.
+This means that, if a function holds a by-object function field, if the type of
+the field is copyable, so too is the function that contains it. This also
+applies to by-copy and by-const-copy captures.
+
+By-reference and by-const-reference captures are a bit different, because even
+if the underlying type is not copyable, the reference to that object is always
+copyable.
+
+The final case is by-value function fields. Since C++ const references, when
+made into fields of a class, prevent the class from being copied, so too should
+by-value function fields prevent the function in which it is contained from
+being copied.
+
+## Self and Recursion
```

Comment:
Given the direction in #3720, an expression of the form `x.(F)`, where `F` is a function with a `self` or `addr self` parameter, produces a callable that holds the value of `x`, and does not hold the value of `F`. As a consequence, I think we can't support combining captures and function fields with a `self` parameter under that model.

Would that be a reasonable restriction to impose here?